File: src/rate_limit_patterns/algorithms/token_bucket.py
"""Token Bucket rate limiting algorithm."""

from __future__ import annotations

import math
from typing import Any

from rate_limit_patterns.models import RateLimitConfig


class TokenBucketAlgorithm:
    """Token Bucket algorithm implementation.

    A token bucket algorithm where tokens are added at a fixed rate
    up to a maximum burst capacity. Each request consumes one token.
    """

    def initial_state(self, config: RateLimitConfig) -> dict[str, Any]:
        """Return initial state with full bucket of tokens.

        Args:
            config: Rate limit configuration.

        Returns:
            Initial state with tokens at burst_size (or limit if no burst).
        """
        burst = config.burst_size if config.burst_size is not None else config.limit
        return {
            "tokens": float(burst),
            "last_refill": 0.0,
        }

    def compute(
        self,
        state: dict[str, Any],
        config: RateLimitConfig,
        current_time: float,
    ) -> tuple[bool, dict[str, Any], dict[str, Any]]:
        """Compute whether a request is allowed and update state.

        Args:
            state: Current algorithm state dictionary.
            config: Rate limit configuration.
            current_time: Current Unix timestamp in seconds.

        Returns:
            Tuple of (allowed: bool, new_state: dict[str, Any], metadata: dict[str, Any]).
        """
        # Ensure pure function - do not mutate incoming state
        last_refill = state["last_refill"]
        current_tokens = state["tokens"]

        # Calculate elapsed time and refill tokens
        elapsed = max(0.0, current_time - last_refill)
        refill_amount = elapsed * config.tokens_per_second

        # Calculate capacity cap
        burst = config.burst_size if config.burst_size is not None else config.limit

        # Calculate new token balance with cap
        new_tokens = min(float(burst), current_tokens + refill_amount)

        # Create new state (pure function - don't mutate input)
        new_state: dict[str, Any] = {
            "tokens": new_tokens,
            "last_refill": current_time,
        }

        # Check if we can consume a token
        if new_tokens >= 1.0:
            # Consume exactly 1 token
            new_state["tokens"] = new_tokens - 1.0
            tokens_after = new_state["tokens"]
            allowed = True
            retry_after: int | None = None
        else:
            # Deny request - calculate retry_after based on fractional tokens needed
            tokens_after = new_tokens
            tokens_needed = 1.0 - tokens_after
            retry_after = math.ceil(tokens_needed / config.tokens_per_second)
            allowed = False

        remaining = max(0, int(math.floor(tokens_after)))
        reset_at = current_time + (burst - tokens_after) / config.tokens_per_second
        request_count = max(0, int(math.floor(burst - tokens_after)))

        meta: dict[str, Any] = {
            "remaining": remaining,
            "reset_at": reset_at,
            "request_count": request_count,
        }
        if not allowed and retry_after is not None:
            meta["retry_after"] = retry_after

        return allowed, new_state, meta

File: src/rate_limit_patterns/backend/lua/token_bucket.lua
-- Token Bucket Rate Limiter
-- KEYS[1] = rate limit key
-- ARGV[1] = burst_size
-- ARGV[2] = tokens_per_second
-- ARGV[3] = current_time (-1 to use Redis TIME)

local key = KEYS[1]
local burst_size = tonumber(ARGV[1])
local tokens_per_second = tonumber(ARGV[2])
local supplied_time = tonumber(ARGV[3])
local now = 0

if supplied_time and supplied_time >= 0 then
    now = supplied_time
else
    local time_data = redis.call('TIME')
    now = tonumber(time_data[1]) + (tonumber(time_data[2]) / 1000000)
end

-- Get current state
local state = redis.call('HMGET', key, 'tokens', 'last_refill')
local tokens = tonumber(state[1]) or burst_size
local last_refill = tonumber(state[2]) or now

-- Calculate token refill
local elapsed = now - last_refill
if elapsed < 0 then
    elapsed = 0
end
local tokens_to_add = elapsed * tokens_per_second
tokens = math.min(tokens + tokens_to_add, burst_size)

-- Check if allowed
local allowed = 0
local retry_after = 0

if tokens >= 1 then
    tokens = tokens - 1
    allowed = 1
else
    local tokens_needed = 1 - tokens
    retry_after = math.ceil(tokens_needed / tokens_per_second)
end

-- Update state
redis.call('HMSET', key, 'tokens', tokens, 'last_refill', now)

local ttl = math.ceil(burst_size / tokens_per_second)
if ttl < 1 then
    ttl = 1
end
redis.call('EXPIRE', key, ttl)

-- Return: allowed, remaining, retry_after, reset_at
local remaining = math.floor(tokens)
local reset_at = now + (burst_size - tokens) / tokens_per_second
local request_count = math.floor(burst_size - tokens)

return {allowed, remaining, retry_after, reset_at, request_count}
